src/flb_pack.c

```diff
@@ -293,3 +293,192 @@ void flb_pack_print(char *data, size_t bytes)
+            ret = FLB_TRUE;
+        }
+        break;
+
+    case MSGPACK_OBJECT_BIN:
+        if (try_to_write(buf,off,left, "\"",1) &&
+            try_to_write(buf,off,left, (char*)o->via.bin.ptr, o->via.bin.size) &&
+            try_to_write(buf,off,left, "\"",1)) {
+            ret = FLB_TRUE;
+        }
+        break;
+
+    case MSGPACK_OBJECT_ARRAY:
+        loop = o->via.array.size;
+
+        if (!try_to_write(buf,off,left, "[",1)) {
+            goto msg2json_end;
+        }
+        if (loop != 0) {
+            msgpack_object* p = o->via.array.ptr;
+            if (!msgpack2json(buf,off,left,p)) {
+                goto msg2json_end;
+            }
+            for (i=1; i<loop; i++) {
+                if (!try_to_write(buf,off,left,", ",2) ||
+                    !msgpack2json(buf,off,left,p+i)) {
+                    goto msg2json_end;
+                }
+            }
+        }
+
+        ret = try_to_write(buf,off,left, "]",1);
+        break;
+
+    case MSGPACK_OBJECT_MAP:
+        loop = o->via.map.size;
+        if (!try_to_write(buf,off,left, "{",1)) {
+            goto msg2json_end;
+        }
+        if (loop != 0) {
+            msgpack_object_kv *p = o->via.map.ptr;
+            if (!msgpack2json(buf,off,left,&p->key) ||
+                !try_to_write(buf,off,left, ":",1)  ||
+                !msgpack2json(buf,off,left,&p->val)) {
+                    goto msg2json_end;
+            }
+            for (i=1; i<loop; i++) {
+                if (
+                  !try_to_write(buf,off,left, ", ",2) ||
+                  !msgpack2json(buf,off,left,&(p+i)->key) ||
+                  !try_to_write(buf,off,left, ":",1)  ||
+                  !msgpack2json(buf,off,left,&(p+i)->val) ) {
+                    goto msg2json_end;
+
+                }
+            }
+        }
+
+        ret = try_to_write(buf,off,left, "}",1);
+        break;
+
+    default:
+        flb_warn("[%s] unknown type",__FUNCTION__);
+    }
+
+ msg2json_end:
+    return ret;
+}
+
+/**
+ *  convert msgpack to JSON string.
+ *  This API is similar to snprintf.
+ *
+ *  @param  json_str The buffer to fill JSON string.
+ *  @param  json_len The size of json_str.
+ *  @param  data     The msgpack_unpacked data.
+ *  @return success  ? a number characters filled : negative value
+ */
+int flb_msgpack_to_json(char* json_str, size_t str_len,
+                        msgpack_unpacked* data)
+{
+    int ret = -1;
+    int off = 0;
+
+    if (json_str == NULL || data == NULL) {
+        return -1;
+    }
+
+    ret = msgpack2json(json_str,&off, str_len, &data->data);
+    json_str[str_len-1] = '\0';
+    return ret ? off: ret;
+}
+
+
+/**
+ *  convert msgpack to JSON string.
+ *  This API is similar to snprintf.
+ *  @param  size     Estimated length of json str.
+ *  @param  data     The msgpack_unpacked data.
+ *  @return success  ? allocated json str ptr : NULL
+ */
+char* flb_msgpack_to_json_str(size_t size, msgpack_unpacked* data)
+{
+    char* ret = NULL;
```

Comment:
most of FLuent Bit code use 'ret' as 'return value' as an intever. For char * please use a different name, e.g: buf, tmp.